MetaMask/metamask-extension
# Fix issue where we show contract address as recipient when calling safe transfer method on erc721 or erc1155 contracts

Fixes: https://github.com/MetaMask/metamask-extension/issues/13855

This fixes a [UX issue that OpenSea raised](https://consensys.slack.com/archives/CN2H44RAM/p1639676146047200) with us a couple months back. IIRC it fell of our radar because after reviewing the behavior it appeared to work. This was a red herring because we derive recipient addresses differently in sends that are initiated within Metamask and those initiated via API. There is indeed still an issue with NFT sends initiated via API:
Current Behavior for in-app NFT send (ignore the image not rendering, separate issue): 

https://user-images.githubusercontent.com/34557516/152865105-41b4bde2-2999-4053-902b-b5e8b74e28fa.mov

Current Behavior for NFT send initialized from OpenSea:

https://user-images.githubusercontent.com/34557516/152865181-40c4f85b-9e89-46e2-a1c4-dd25d9a5be6c.mov

With this fix:

https://user-images.githubusercontent.com/34557516/152865686-1b4b5221-764f-496e-98d8-b67ff0108373.mov

- Moves `getTransactionData` method to a shared module and make use of it in the method `determineTransactionType` already in that shared module.

Manual testing steps:  
  - Attempt to transfer an NFT via OpenSea (or Etherscan)
  - Make sure that the "recipient address" in the upper right hand corner matches the recipient you entered, rather than the contract address.
  
Next Steps:
- I opened this ticket to migrate our ABI usage towards the pattern that is used here (importing and making use of all of our known/standard ABIs - ERC20, ERC721, ERC1155). Alternatively we may want to expose methods on AssetsContractController and only instantiate the ABI interfaces there.
